File: src/main/java/com/gorogoro/auth/user/adapter/in/web/UserController.java
package com.gorogoro.auth.user.adapter.in.web;

import com.gorogoro.auth.user.adapter.in.web.request.RegisterUserRequest;
import com.gorogoro.auth.user.adapter.in.web.request.UpdateUserRequest;
import com.gorogoro.auth.user.adapter.in.web.response.UserResponse;
import com.gorogoro.auth.user.application.dto.command.RegisterUserCommand;
import com.gorogoro.auth.user.application.dto.command.UpdateUserCommand;
import com.gorogoro.auth.user.application.port.in.GetUserUseCase;
import com.gorogoro.auth.user.application.port.in.RegisterUserUseCase;
import com.gorogoro.auth.user.application.port.in.UpdateUserUseCase;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PatchMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequestMapping("/api/users")
@RequiredArgsConstructor
public class UserController {
    private final RegisterUserUseCase registerUserUseCase;
    private final GetUserUseCase getUserUseCase;
    private final UpdateUserUseCase updateUserUseCase;

    @PostMapping("/register")
    public ResponseEntity<UserResponse> register(@Valid @RequestBody RegisterUserRequest request) {
        RegisterUserCommand command = RegisterUserCommand.toCommand(request);
        return ResponseEntity.status(HttpStatus.CREATED)
                .body(UserResponse.from(registerUserUseCase.register(command)));
    }

    @GetMapping("/info")
    public ResponseEntity<UserResponse> getUser(@AuthenticationPrincipal UserDetails userDetails) {
        return ResponseEntity.ok(
                UserResponse.from(getUserUseCase.getUser(Long.valueOf(userDetails.getUsername())))
        );
    }

    @PatchMapping("/update")
    public ResponseEntity<UserResponse> updateUser(
            @AuthenticationPrincipal UserDetails userDetails,
            @Valid @RequestBody UpdateUserRequest request
    ) {
        UpdateUserCommand command = UpdateUserCommand.of(Long.valueOf(userDetails.getUsername()), request);
        return ResponseEntity.ok(
                UserResponse.from(updateUserUseCase.updateUser(command))
        );
    }
}

File: src/main/kotlin/com/gorogoro/auth/user/infra/adapter/in/web/UserController.kt
package com.gorogoro.auth.user.infra.adapter.`in`.web

import com.gorogoro.auth.user.application.dto.GetUserCommand
import com.gorogoro.auth.user.application.dto.UpdateUserCommand
import com.gorogoro.auth.user.application.port.`in`.GetUserUseCase
import com.gorogoro.auth.user.application.port.`in`.ModifyUserUseCase
import com.gorogoro.auth.user.infra.adapter.`in`.web.request.UpdateUserRequest
import com.gorogoro.auth.user.infra.adapter.`in`.web.response.FindUserResponse
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PatchMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestHeader
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("/api/users")
class UserController(
    private val modifyUserUseCase: ModifyUserUseCase,
    private val getUserUseCase: GetUserUseCase,
) {
    @PatchMapping("/modify")
    fun updateUser(
        @RequestHeader("X-User-Id") userId: Long,
        @RequestBody request: UpdateUserRequest
    ): ResponseEntity<Void> {
        val command = UpdateUserCommand(
            userId = userId,
            name = request.name,
            nickname = request.nickname,
            email = request.email,
            passwordEncrypted = request.passwordEncrypted
        )

        modifyUserUseCase.updateUserInfo(command)
        return ResponseEntity.ok().build()
    }

    @DeleteMapping("/leave")
    fun withdrawUser(
        @RequestHeader("X-User-Id") userId: Long,
    ): ResponseEntity<Void> {
        modifyUserUseCase.leaveUser(userId)
        return ResponseEntity.ok().build()
    }

    @GetMapping("/info")
    fun getUserInfo(@RequestHeader("X-User-Id") userId: Long): ResponseEntity<FindUserResponse> {
        val user = getUserUseCase.getUserInfo(GetUserCommand(userId))
        return ResponseEntity.ok(user)
    }
}
